V7: Documentation (#259)
Financial-Times/o-forms: README.md

```html
...
	<span class="o-forms-input o-forms-input--radio-box">
		<div class="o-forms-input--radio-box__container">
			<label>
				<input type="radio" name="my-box-radio"/>
				<span class="o-forms-input__label">Radio 1</span>
			</label>
			<label>
				<input type="radio" name="my-box-radio"/>
				<span class="o-forms-input__label">Radio 2</span>
			</label>
		</div>
	</span>
...
```

This input type also accepts a 'negative' modifier `o-forms-input__label--negative`, which changes the background colour of an input when selected (usually for a negative choice):
```html
<div class="o-forms-field">
	...
	<span class="o-forms-input o-forms-input--radio-box">
		<div class="o-forms-input--radio-box__container">
			<label>
				<input type="radio" name="negative" value="Yes">
				<span class="o-forms-input__label">Yes</span>
			</label>
			<label>
				<input type="radio" name="negative" value="No"checked>
				<span class="o-forms-input__label o-forms-input__label--negative">No</span>
			</label>
		</div>
	</span>
</div>
```
[_See the full markup for a box-style radio button in the registry_](https://registry.origami.ft.com/components/o-forms#radio-box)

> #### `input[type=radio]`

Reviewer: Looking at naming as well, but just spotted this
```suggestion
#### `input[type=checkbox]`
```

</div>
```
[_See the full markup for an anchor in the registry_](https://registry.origami.ft.com/components/o-forms#anchors)

### Shared Modifiers
The markup structure outlined at the beginning of this section indicates containers for a form field that accept modifiers.

#### Field Modifiers
The following modifiers are available for the `o-forms-field` element:
- `o-forms-field--optional`: indicates that a field is optional, and will append `(optional)` to the end of the field's title.
- `o-forms-field--inline`: inlines the title container to the input container

#### Input Modifiers
- `o-forms-input--valid`: identifies a valid input, gets set dynamically on form valdiation
- `o-forms-input--invalid`: identifies an invalid input, gets set dynamically on form valdiation. This requires additional markup to be added as a direct child of the `o-forms-input` element:
```diff
-<span class="o-forms-input">
+<span class="o-forms-input o-forms--invalid">
	<!-- input(s) -->
+	<span class="o-forms-input__error>Error message here</span>
</span>
```
- `o-forms-input--inline`: inlines all inputs with each other (usually only used on multiple input fields).

### Errors

> In trms of invalid inputs, `o-forms` has a build in mechanism to display custom errors and an error summary. If neither of these work for your product, you can also default to [native browser validation](#form-instance).

Reviewer 2: ```suggestion
In terms of invalid inputs, `o-forms` has a build in mechanism to display custom errors and an error summary. If neither of these work for your product, you can also default to [native browser validation](#form-instance).
```

#### Custom Errors
In order to provide customised error messages for an invalid input field, you'll need to add the message to the markup. In any of the structures shown in the markup above, the error message will need to live within the input container as a sibling to the input elements:
```html
...
	<span class="o-forms-input">
		<!-- inputs -->
		<span class="o-forms-input__error">This is the error message</span>
	</span>
...
```
The message is hidden by default, until the input field become invalid. 

#### Error Summary
`o-forms` also generates an error message element when a form is submitted and invalid inputs are recognised.

This feature will collect the custom messages of the invalid fields if they are present in the markup, or will default to the browsers native error message if they aren't. It will associate the message to the title of the input that is invalid, generate a list of links at the top of the form, and focus on the first link.

> This markup will always generated dynamically if the [errorSummary option](#form-instance) is set to `true`.

Reviewer 2: ```suggestion
This markup will always be generated dynamically if the [errorSummary option](#form-instance) is set to `true`.
```

## Sass

> You can include styles for every input type and feature by calling the `o-forms` primary mixin:

Reviewer 2: I may be misunderstanding, but I think the primary mixin is called `oForms` and the sass file to import is called `o-forms`. 
I'm not sure if my suggestion is correct though so feel free to resolve this without changing it.
```suggestion
You can include styles for every input type and feature by calling the `oForms` primary mixin:
```

Author: Oh it was meant to imply the primary mixin that belongs to `o-forms` but yeah now that you point it out it is kind of confusing 🤔

```scss
@import 'o-forms';
@include oForms();
```

If you would like to be more specific about what aspects of the styles get output, then you'll need to provide the primary mixin with an `$opts` map. For example, in order to output styles for text inputs and checkboxes, and add the abilty to inlne them, you would use the following:
```scss
@import 'o-forms';
@include oForms($opts: (
	'elements': ('text-input', 'checkbox'),
	'features': ('inline')
));
```
### Options
`o-forms` has many options due to its comprehensive nature.  
The `$opts` map accepts two lists with the following options:
- `'elements'`:
	- `'anchor'`
	- `'checkbox'`
	- `'date'`
	- `'password'`
	- `'radio-round'`
	- `'radio-box'`
	- `'select'`
	- `'textarea'`
	- `'text'`
	- `'toggle'`
- `'features'`:
	- `'disabled'`
	- `'inline'`
	- `'right'`
	- `'inverse'`
	- `'small'`
	- `'state'`
	- `'suffix'`

> _Note 1: `valid` and `invalid` styles are output with the base of `o-forms` so there is no need to include them in the list above._

Reviewer 2: 👍

> _Note 2: It's important to remember that not all features will apply to all input types. If there are any features you would like to see added to an input, please [get in touch with the team](#contact)._

Reviewer 2: 👍
